keps/sig-network/0034-20190125-ingress-api-group.md

```diff
@@ -0,0 +1,102 @@
+---
+kep-number: 35
+title: Move Ingress to the networking.v1beta1 API group
+authors:
+  - "@bowei"
+owning-sig: sig-network
+participating-sigs:
+reviewers:
+  - "@alefbd"
+approvers:
+  - "@thockin"
+  - "@caseydavenport"
+editor:
+creation-date: 2018-01-25
+last-updated: 2018-01-27
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# Move Ingress to the networking.v1beta1 API group
+
+## Summary
+
+Copy the Ingress resource from the current API group (extensions.v1beta1) to networking.v1beta1 and prepare the resource to go to GA as-is as the Ingress.V1.
+
+## Motivation
+
+The `extensions` API group is considered deprecated.  Ingress is the last non-deprecated API in that group.  All other types have been migrated to other permanent API groups.  Such an API group migration takes three minor version cycles (~9 months) to ensure compatibility. This means any API group movement should be started sooner rather than later.
+
+The Ingress resource has been in a beta state for a *long* time (first commit was in Fall 2015). While the interface [is not perfect](https://github.com/bowei/k8s-ingress-survey-2018), there are many [independent implementations](https://kubernetes.io/docs/concepts/services-networking/ingress/#ingress-controllers) in active use.
+
+We have a couple of choices (and non-choices, see appendix) for the current resource:
+
+1.  We can delete the current resource from extensions.v1beta1 in anticipation that an improved API can replace it.
+1.  We can copy the API as-is (or with minor changes) into networking.v1beta1, preserving/converting existing data (following the same approach taken with all other extensions.v1beta1 resources). This will allow us to start the cleanup of the extensions API group. This also prepares the API for GA.
+
```

Comment:
You can also move it to a CRD in its own group, with either a GA move or simply leaving it In permanent beta (avoiding the churn on version).

Reply:
I asked this, but Jordan said there's no way to convert existing Ingress users to a CRD transparently, which makes sense.  We even wrote it down in the appendix

Reply:
In the past, leaving APIs in limbo has caused user confusion and concern, so I don't think permanent beta is a good option